File: app/src/main/java/me/smartproxy/dns/DnsHeader.java
package me.smartproxy.dns;

import java.nio.ByteBuffer;

import me.smartproxy.tcpip.CommonMethods;


public class DnsHeader {
    public short ID;
    public DnsFlags Flags;
    public short QuestionCount;
    public short ResourceCount;
    public short AResourceCount;
    public short EResourceCount;

    public static DnsHeader fromBytes(ByteBuffer buffer) {
        DnsHeader header = new DnsHeader(buffer.array(), buffer.arrayOffset() + buffer.position());
        header.ID = buffer.getShort();
        header.Flags = DnsFlagsKt.parse(buffer.getShort());
        header.QuestionCount = buffer.getShort();
        header.ResourceCount = buffer.getShort();
        header.AResourceCount = buffer.getShort();
        header.EResourceCount = buffer.getShort();
        return header;
    }

    public void toBytes(ByteBuffer buffer) {
        buffer.putShort(this.ID);
        buffer.putShort(this.Flags.toShort());
        buffer.putShort(this.QuestionCount);
        buffer.putShort(this.ResourceCount);
        buffer.putShort(this.AResourceCount);
        buffer.putShort(this.EResourceCount);
    }

    private static final short offset_ID = 0;
    private static final short offset_Flags = 2;
    private static final short offset_QuestionCount = 4;
    private static final short offset_ResourceCount = 6;
    private static final short offset_AResourceCount = 8;
    private static final short offset_EResourceCount = 10;

    public byte[] Data;
    public int Offset;

    public DnsHeader(byte[] data, int offset) {
        this.Offset = offset;
        this.Data = data;
    }

    public short getID() {
        return CommonMethods.readShort(Data, Offset + offset_ID);
    }

    public short getFlags() {
        return CommonMethods.readShort(Data, Offset + offset_Flags);
    }

    public short getQuestionCount() {
        return CommonMethods.readShort(Data, Offset + offset_QuestionCount);
    }

    public short getResourceCount() {
        return CommonMethods.readShort(Data, Offset + offset_ResourceCount);
    }

    public short getAResourceCount() {
        return CommonMethods.readShort(Data, Offset + offset_AResourceCount);
    }

    public short getEResourceCount() {
        return CommonMethods.readShort(Data, Offset + offset_EResourceCount);
    }

    public void setID(short value) {
        CommonMethods.writeShort(Data, Offset + offset_ID, value);
    }

    public void setFlags(short value) {
        CommonMethods.writeShort(Data, Offset + offset_Flags, value);
    }

    public void setQuestionCount(short value) {
        CommonMethods.writeShort(Data, Offset + offset_QuestionCount, value);
    }

    public void setResourceCount(short value) {
        CommonMethods.writeShort(Data, Offset + offset_ResourceCount, value);
    }

    public void setAResourceCount(short value) {
        CommonMethods.writeShort(Data, Offset + offset_AResourceCount, value);
    }

    public void setEResourceCount(short value) {
        CommonMethods.writeShort(Data, Offset + offset_EResourceCount, value);
    }
}

File: app/src/main/java/me/smartproxy/dns/DnsHeader.kt
package me.smartproxy.dns

import me.smartproxy.core.Pool
import me.smartproxy.tcpip.CommonMethods
import java.nio.ByteBuffer

class DnsHeader {
    lateinit var data: ByteArray

    var offset: Int = 0

    var ID: Short
        set(value) {
            CommonMethods.writeShort(data, offset + offset_ID, value)
        }
        get() {
            return CommonMethods.readShort(data, offset + offset_ID)
        }
    lateinit var Flags: DnsFlags

    var questionCount: Short
        set(value) {
            CommonMethods.writeShort(data, offset + offset_QuestionCount, value)
        }
        get() {
            return CommonMethods.readShort(data, offset + offset_QuestionCount)
        }

    var resourceCount: Short
        set(value) {
            CommonMethods.writeShort(data, offset + offset_ResourceCount, value)
        }
        get() {
            return CommonMethods.readShort(data, offset + offset_ResourceCount)
        }

    var aResourceCount: Short
        set(value) {
            CommonMethods.writeShort(data, offset + offset_AResourceCount, value)
        }
        get() {
            return CommonMethods.readShort(data, offset + offset_AResourceCount)
        }

    var eResourceCount: Short
        set(value) {
            CommonMethods.writeShort(data, offset + offset_EResourceCount, value)
        }
        get() {
            return CommonMethods.readShort(data, offset + offset_EResourceCount)
        }

    fun toBytes(buffer: ByteBuffer) {
        buffer.putShort(this.ID)
        buffer.putShort(Flags.toShort())
        buffer.putShort(this.questionCount)
        buffer.putShort(this.resourceCount)
        buffer.putShort(this.aResourceCount)
        buffer.putShort(this.eResourceCount)
    }


    companion object {
        fun recycle(header: DnsHeader) {
            DnsHeaderPool.recycle(header)
        }

        fun takeFromPoll(buffer: ByteBuffer): DnsHeader {
            val header = DnsHeaderPool.take()
            header.data = buffer.array()
            header.offset = buffer.arrayOffset() + buffer.position()
            header.ID = buffer.getShort()
            header.Flags = parse(buffer.getShort())
            header.questionCount = buffer.getShort()
            header.resourceCount = buffer.getShort()
            header.aResourceCount = buffer.getShort()
            header.eResourceCount = buffer.getShort()
            return header
        }

        private const val offset_ID: Short = 0
        private const val offset_Flags: Short = 2
        private const val offset_QuestionCount: Short = 4
        private const val offset_ResourceCount: Short = 6
        private const val offset_AResourceCount: Short = 8
        private const val offset_EResourceCount: Short = 10
    }
}

object DnsHeaderPool : Pool<DnsHeader>() {
    override fun create(): DnsHeader {
        return DnsHeader()
    }
}